### PR TITLE
Provide basic NIO-based PathUtils (#98)

### PR DESCRIPTION
**Pull request by boris-unckel**
Fixes https://github.com/smallrye/smallrye-common/issues/98

### CODE REVIEW

**Comment by boris-unckel** [comment]
Hello @dmlloyd as you suggested I have prepared the NIO utils PR here in smallrye-common. Please have a look and review.

File: io/src/main/java/io/smallrye/common/io/FileUtils.java [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+     * {@link java.nio.file.Files#isDirectory(Path, LinkOption...)} but with
+     * explicit null check and unchecked {@code java.io.UncheckedIOException}.
+     *
+     * @param file to be checked, must be not null
+     * @return the passed path if it is a regular file
+     * @throws IllegalArgumentException if the path is {@code null}
+     * @throws UncheckedIOException if is not a regular file or determination of
+     *         type fails
+     */
+    public static File checkIsRegularFileNoFollowLinks(final File file) throws UncheckedIOException {
```
Comment (dmlloyd) [inline io/src/main/java/io/smallrye/common/io/FileUtils.java:362]: I noticed that some methods take `*Options*` flags and some do not (i.e. `CopyOptions` vs `LinkOptions`); is there a reason for the inconsistency?

File: io/src/main/java/io/smallrye/common/io/FileUtils.java [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+     * @throws UncheckedIOException if directory contains elements or check
+     *         fails
+     */
+    public static File checkIsEmptyDirectory(final File dir) throws UncheckedIOException {
+        checkNotNullParam("dir", dir);
+        return PathUtils.doCheckIsEmptyDirectory(dir.toPath()).toFile();
+    }
+
+    private FileUtils() {
+        throw Assert.unreachableCode();
```
Comment (dmlloyd) [inline io/src/main/java/io/smallrye/common/io/FileUtils.java:562]: It is enough to have the constructor be private IMO.  This is consistent with other classes in the project I believe.

File: io/src/main/java/io/smallrye/common/io/PathUtils.java [hunk 3]
```diff
@@ -0,0 +1,2139 @@
...
+     * A wrapper around {@link java.nio.file.Paths#get(String, String...)} with
+     * explicit null check.
+     *
+     * @param first part of the path, must be not null, not empty and a valid path
+     * @param more optional parts of the path
+     * @return the path constructed out of first and more
+     * @throws IllegalArgumentException if first is null or empty
+     */
+    public static Path get(final String first, final String... more) {
+        return Paths.get(checkNotEmptyAfterTrimParam("first", first), more);
```
Comment (dmlloyd) [inline io/src/main/java/io/smallrye/common/io/PathUtils.java:55]: Same comment as above.

File: io/src/main/java/io/smallrye/common/io/PathUtils.java [hunk 4]
```diff
@@ -0,0 +1,2139 @@
...
+    /**
+     * A wrapper around {@link java.nio.file.Paths#get(URI)} with explicit null
+     * check.
+     *
+     * @param uri for the path, must be not null and a valid path
+     * @return the path constructed out of the uri
+     * @throws IllegalArgumentException if uri is null
+     */
+    public static Path get(final URI uri) {
+        return Paths.get(checkNotNullParam("uri", uri));
```
Comment (dmlloyd) [inline io/src/main/java/io/smallrye/common/io/PathUtils.java:67]: I believe that `Paths.get()` also does a `null` check; why wrap it?

File: io/src/main/java/io/smallrye/common/io/PathUtils.java [hunk 5]
```diff
@@ -0,0 +1,2139 @@
...
+    /**
+     * A wrapper around {@link java.nio.file.Files#newDirectoryStream(Path)} with
+     * explicit null check and unchecked {@code java.io.UncheckedIOException}.
+     *
+     * @param dir to the directory to list, must be not null.
+     * @return the elements in the directory as DirectoryStream
+     * @throws IllegalArgumentException if the path is {@code null}
+     * @throws UncheckedIOException around original IOExceptions
+     */
+    public static DirectoryStream<Path> newDirectoryStream(final Path dir) throws UncheckedIOException {
```
Comment (dmlloyd) [inline io/src/main/java/io/smallrye/common/io/PathUtils.java:79]: I personally don't much like wrapping with `UncheckedIOException`.  We don't do that in just about any other place that I can think of. How would you justify this feature?

File: io/src/main/java/io/smallrye/common/io/PathUtils.java [hunk 6]
```diff
@@ -0,0 +1,2139 @@
...
+            }
+            throw new UncheckedIOException(deleteFailed);
+        }
+        return path;
+    }
+
+    /**
+     * Recursive deletion of directories content, including subfolders.
+     *
+     * Adopted from Google Guava.
```
Comment (dmlloyd) [inline io/src/main/java/io/smallrye/common/io/PathUtils.java:751]: Is there a license concern here?

File: io/src/main/java/io/smallrye/common/io/PathUtils.java [hunk 7]
```diff
@@ -0,0 +1,2139 @@
...
+     *
+     * @param path to delete, must be not null, a path, and exists.
+     * @return the path to delete
+     * @throws UncheckedIOException around original IOExceptions and if not exists
+     *         path
+     * @see <a href=
```
Comment (dmlloyd) [inline io/src/main/java/io/smallrye/common/io/PathUtils.java:757]: The content of `@see` is a member reference; it's not intended for external references.

File: io/src/main/java/io/smallrye/common/io/PathUtils.java [hunk 8]
```diff
@@ -0,0 +1,2139 @@
...
+                deleteFailed.addSuppressed(e);
+            }
+            throw new UncheckedIOException(deleteFailed);
+        }
+        return path;
+    }
+
+    /**
+     * Recursive deletion of directories content, including subfolders.
+     *
```
Comment (dmlloyd) [inline io/src/main/java/io/smallrye/common/io/PathUtils.java:750]: You should put a `<p>` when you want to separate doc paragraphs.

File: io/src/main/java/io/smallrye/common/io/PathUtils.java [hunk 9]
```diff
@@ -0,0 +1,2139 @@
...
+     *         path
+     * @see <a href=
+     *      "https://github.com/google/guava/blob/master/guava/src/com/google/common/io/MoreFiles.java">com.google.common.io.MoreFiles</a>
+     */
+    public static Path deleteDirectoryContents(Path path) throws UncheckedIOException {
+        return doDeleteDirectoryContents(checkIsDirectoryNoFollowLinks(path));
+    }
+
+    static Path doDeleteDirectoryContents(Path path) throws UncheckedIOException {
+        Collection<IOException> exceptions = null; // created lazily if needed
```
Comment (dmlloyd) [inline io/src/main/java/io/smallrye/common/io/PathUtils.java:765]: This is minor but you could have this variable be of type `FileSystemException` and pass that in to the nested methods rather than adding the suppressed exceptions to a collection only to iterate them all right away.

File: io/src/main/java/io/smallrye/common/io/FileUtils.java [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+     * A wrapper around {@link java.nio.file.Paths#get(String, String...)} with
+     * explicit null check.
+     *
+     * @param first part of the path, must be not null, not empty and a valid path
+     * @param more optional parts of the path
+     * @return the path constructed out of first and more
+     * @throws IllegalArgumentException if first is null or empty
+     */
+    public static File get(final String first, final String... more) {
+        return Paths.get(checkNotEmptyAfterTrimParam("first", first), more).toFile();
```
[the remark this answers is not in the document]
Comment (boris-unckel) [reply io/src/main/java/io/smallrye/common/io/FileUtils.java:40]: Thanks for your valuable time to review, I really appreciate that.

- Trimming whitespace at all: On both Windows and Linux/Unixiod systems path elements with leading or trailing space cause trouble. The method does not clean whitespace inside the names. The freedom to use leading/trailing whitespace is still there, but the default tries to avoid it. This is only for Strings the case, all other objects (Path, File, URI) allow it. I have documented the reason with links to detailed documentation in the class javadoc.
- Consistency between "first" and "more" parameter: thanks for the catch, I've fixed it.

File: io/src/main/java/io/smallrye/common/io/FileUtils.java (hunk 1 above)
Comment (boris-unckel) [reply io/src/main/java/io/smallrye/common/io/FileUtils.java:362]: Yes, this is intended. (No)FollowLinks is important to understand erroneous behaviour, this makes it directly visible in stackstraces and the developer will care, because even autocompletion shows directly the alternatives.
The copy options with it's multiple combinations would blow the code, I assume someone who is looking for "replace" instead of getting a "FileAlreadyExistsException" will start to look at the docs or google stackoverflow.
I have documented the approach in the class javadoc, including links.

File: io/src/main/java/io/smallrye/common/io/FileUtils.java (hunk 2 above)
Comment (boris-unckel) [reply io/src/main/java/io/smallrye/common/io/FileUtils.java:562]: Removed.

File: io/src/main/java/io/smallrye/common/io/PathUtils.java (hunk 3 above)
Comment (boris-unckel) [reply io/src/main/java/io/smallrye/common/io/PathUtils.java:55]: Same fix.

File: io/src/main/java/io/smallrye/common/io/PathUtils.java (hunk 4 above)
Comment (boris-unckel) [reply io/src/main/java/io/smallrye/common/io/PathUtils.java:67]: I have checked that and could not find a explicit null check. The intention is to have a Exception with the parameter name. I don't have a personal favour for NullPointer vs. IllegalArgument but I dislike the anonymous, fail late ones.

File: io/src/main/java/io/smallrye/common/io/PathUtils.java (hunk 7 above)
Comment (boris-unckel) [reply io/src/main/java/io/smallrye/common/io/PathUtils.java:757]: Fixed, moved to text.

File: io/src/main/java/io/smallrye/common/io/PathUtils.java (hunk 8 above)
Comment (boris-unckel) [reply io/src/main/java/io/smallrye/common/io/PathUtils.java:750]: Fixed.

File: io/src/main/java/io/smallrye/common/io/PathUtils.java (hunk 6 above)
Comment (boris-unckel) [reply io/src/main/java/io/smallrye/common/io/PathUtils.java:751]: No, Google Guava is licensed under the Apache v2 license. [https://github.com/google/guava/blob/master/COPYING](https://github.com/google/guava/blob/master/COPYING)

File: io/src/main/java/io/smallrye/common/io/PathUtils.java (hunk 5 above)
Comment (boris-unckel) [reply io/src/main/java/io/smallrye/common/io/PathUtils.java:79]: Before I started with any of the coding I have searched for places which might be worth to be refactored from classic java.io to java.nio. Many follow a pattern with private utility methods declaring a throws IOException (without Exception handling) and a major place where no Exception handling logic appears but to rethrow it as unchecked Exception.
You're right that currently the usage is different. I think the basic reason is backward compatibility of the JDK and missing 3rd party libraries supporting it.

I have added a section for this in the class javadoc.

File: io/src/main/java/io/smallrye/common/io/PathUtils.java (hunk 9 above)
Comment (boris-unckel) [reply io/src/main/java/io/smallrye/common/io/PathUtils.java:765]: The current approach allows lazy init. If I change that, the top-most exception would have a stacktrace top of a utility method or I can't lazy init.

File: io/src/main/java/io/smallrye/common/io/PathUtils.java (hunk 5 above)
Comment (dmlloyd) [reply io/src/main/java/io/smallrye/common/io/PathUtils.java:134]: I think the feature would *ideally* have to be justifiable within the context of SmallRye projects, or else be widely necessary by consumers of SmallRye projects, which is why I asked.  Unchecked versions of things is a specific enough need that I would imagine it to be framed specifically that way: e.g. `UncheckedPaths`, `UncheckedFiles` etc.  Also, specifically for methods like this one which return an object which is a subtype of `Closeable`, I'd expect a wrapper object which throws an unchecked exception on close e.g. `interface UncheckedDirectoryStream<Path> extends DirectoryStream<Path> { void close() throws UncheckedIOException; }`.

Put more simply, making unchecked versions of NIO methods is a separate concern from formalizing `null` checks according to a different scheme (though it would certainly be reasonable to perform `null` checks according to that scheme from such an API), and I'd expect the API to reflect that.

Comment (boris-unckel) [reply io/src/main/java/io/smallrye/common/io/PathUtils.java:134]: Yes, something I didn't think of: It's not enough to provide "unchecked" utility methods, the underlying objects should behave in the same way.
This leads to a new problem: If UncheckedPath extends Path it's possible to provide different exception signatures, but a wrapper implementation can only throw either checked xor unchecked exceptions, not depending on a cast. All other objects (channels, buffer...) would still require checked ex behaviour.

In short: this is not ready to launch, neither in total nor in parts.

**Comment by boris-unckel** [comment]
@dmlloyd Before I discard this PR I wanted to say thank you again for investing your valuable time. I really appreciate that. I have learned a lot,  about class design and an open source process.
